### PR TITLE
feat(encodable): implement axis functions for ChannelEncoder

### PR DESCRIPTION
**Pull request by kristw**
🏆 Enhancements

* Add function for getting axis title, tick labels.

### CODE REVIEW

File: packages/superset-ui-encodable/src/encoders/ChannelEncoder.ts [hunk 1]
```diff
@@ -1,7 +1,12 @@
 import { extent as d3Extent } from 'd3-array';
+import { HasToString, IdentityFunction } from '../types/Base';
 import { ChannelType, ChannelInput } from '../types/Channel';
 import { PlainObject, Dataset } from '../types/Data';
 import { ChannelDef } from '../types/ChannelDef';
+import { Value } from '../types/VegaLite';
```
Comment (kristw) [inline packages/superset-ui-encodable/src/encoders/ChannelEncoder.ts]: Just change the order

File: packages/superset-ui-encodable/src/encoders/ChannelEncoderAxis.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      this.config.format || '',
+    );
+  }
+
+  getTitle() {
+    return this.config.title;
+  }
+
+  hasTitle() {
+    return this.config.title !== '';
```
Comment (williaster) [inline packages/superset-ui-encodable/src/encoders/ChannelEncoderAxis.ts]: I can't quite tell from the PR, can `title` be `undefined` or `null`? could do `!!this.config.title`

File: packages/superset-ui-encodable/src/encoders/ChannelEncoderAxis.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  }
+
+  getTickLabels() {
+    const { tickCount, values } = this.config;
+
+    if (typeof values !== 'undefined') {
+      return (values as (
+        | number
```
Comment (williaster) [inline packages/superset-ui-encodable/src/encoders/ChannelEncoderAxis.ts]: could make a separate `type` (or re-use the `config.values` type?)

File: packages/superset-ui-encodable/src/encoders/ChannelEncoderAxis.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        | boolean
+        | DateTime
+      )[]).map((v: number | string | boolean | DateTime) =>
+        this.formatValue(isDateTime(v) ? parseDateTime(v) : v),
+      );
+    }
+
+    const { scale } = this.channelEncoder;
+    if (scale && 'domain' in scale) {
+      return ('ticks' in scale ? scale.ticks(tickCount) : scale.domain()).map(this.formatValue);
```
Comment (williaster) [inline packages/superset-ui-encodable/src/encoders/ChannelEncoderAxis.ts]: is the `domain` check relevant for the first part of the ternary where `ticks` is used instead?

File: packages/superset-ui-encodable/test/encoders/ChannelEncoderAxis.test.ts [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+          type: 'quantitative',
+          field: 'speed',
+        },
+      });
+      expect(encoder.axis).toBeDefined();
+    });
+  });
+
+  describe('.formatValue()', () => {
+    it('format value', () => {
```
Comment (williaster) [inline packages/superset-ui-encodable/test/encoders/ChannelEncoderAxis.test.ts]: nit – "format" => "formats"

File: packages/superset-ui-encodable/test/encoders/ChannelEncoderAxis.test.ts [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+          type: 'quantitative',
+          field: 'speed',
+          axis: {
+            format: '.2f',
+          },
+        },
+      });
+      expect(encoder.axis && encoder.axis.formatValue(200)).toEqual('200.00');
+    });
+    it('fallbacks to field formatter', () => {
```
Comment (williaster) [inline packages/superset-ui-encodable/test/encoders/ChannelEncoderAxis.test.ts]: nit – "fallbacks" => "fallsback"

File: packages/superset-ui-encodable/test/encoders/ChannelEncoderAxis.test.ts [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+        channelType: 'X',
+        definition: {
+          type: 'quantitative',
+          field: 'speed',
+          format: '.3f',
+        },
+      });
+      expect(encoder.axis && encoder.axis.formatValue(200)).toEqual('200.000');
+    });
+    it('fallbacks to default formatter', () => {
```
Comment (williaster) [inline packages/superset-ui-encodable/test/encoders/ChannelEncoderAxis.test.ts]: nit – "fallbacks" => "fallsback"

File: packages/superset-ui-encodable/src/encoders/ChannelEncoderAxis.ts (hunk 3 above)
Comment (kristw) [reply packages/superset-ui-encodable/src/encoders/ChannelEncoderAxis.ts]: updated to use util `inferElementTypeFromUnionOfArrayTypes`

File: packages/superset-ui-encodable/src/encoders/ChannelEncoderAxis.ts (hunk 2 above)
Comment (kristw) [reply packages/superset-ui-encodable/src/encoders/ChannelEncoderAxis.ts]: This is `CompleteAxisConfig` which guarantees `title` will be `string`.

File: packages/superset-ui-encodable/src/encoders/ChannelEncoderAxis.ts (hunk 4 above)
Comment (kristw) [reply packages/superset-ui-encodable/src/encoders/ChannelEncoderAxis.ts]: it turns out to be relevant because it triggers some type inference.

**Pull request merged**